p4-16/psa/PSA.mdk

```diff
@@ -477,8 +477,7 @@ The list above is not intended to be exhaustive.
 All of the comments above apply to all types where numerical
 translation occurs between the controller and the data plane.
 Below is a complete list of such types planned for numerical
-translation by default with PSA, and have an annotation
-`p4runtime_translation` in the `psa.p4` include file:
+translation by default with PSA:
```

Comment:
so it seems that the term "numerical translation by default" implies that the server must support any bijective function between the set of SDN ports and the set of device ports?
for types with no default numerical translation (all PSA types save for `PortId_t` and `ClassOfService_t`), the server is expected to at least support type casting between the SDN type and the device type (truncation or padding as appropriate)?

Reply:
I am open to clearer wording here if you have any ideas.  Originally when this text was written, we were only planning on `PortId_t` and `ClassOfService_t` having numerical translation on, and it seemed highly expected that for `PortId_t` that pretty much all PSA implementations would be doing so.

What do you think of changing that last sentence to "Below is a complete list of such planned types where many P4Runtime agents will perform numerical translation in PSA:"

Reply:
I don't really have a good suggestion TBH, I wast just checking that we are on the same page. By adding a `p4runtime_translation` annotation to every type, we are basically implying that translation could be expected for all the types. When in reality, we are just using `p4runtime_translation` to suggest that the P4Info bitwidth should be the same regardless of the device. For all types except `PorId_t` and `ClassOfService_t`, we expect that the "translation" will likely be a simple cast.

I think the rest of your changes make this clear, which is why I approved the PR:

> Note that all of these types have a `@p4runtime_translation`
annotation in the `psa.p4` include file.  This is to ensure that when
the compiler generates a P4Runtime P4Info file from source programs,
it will not include the target-specific bit width in the P4Info file.
For the same P4 source program, the P4Info file contents are intended
to be identical for all targets.
This also enables P4Runtime systems to easily enable numerical
translation for any of these types later, if desired.

So to summarize it seems that the presence `@p4runtime_translation` implies that the P4Runtime server may have to perform casting between the SDN type and the device type. Having the annotation also implies that more complex "numerical translation" (arbitrary bijection) can be enabled at runtime. This is done by associating the translation URI with a specific bijection and communicating this to the P4Runtime server. We expect that a P4Runtime server running on a PSA device should be able to support arbitrary "numerical translation" for `PorId_t` and `ClassOfService_t` (other types are optional and only need to support basic casting).

If we are on the same page, I would just clarify the above block of text I quote with: "If the type bitwidth specified as the second parameter to the `p4runtime_annotation` is different from the device-specific bitwidth (of the underlying type), we expect the P4Runtime server to perform the appropriate casting. Additionally more advance numerical translation can be enabled at runtime for any type annotated with `p4runtime_translation`, although arbitrary numerical translation is only mandated for `PorId_t` and `ClassOfService_t`. To request arbitrary numerical translation for a give type, the P4Runtime system will expect the URI (first parameter to the  `p4runtime_annotation`) and the desired mapping."

Reply:
We are on the same page.  I added the text you suggested, in place of the last sentence of mine that you quoted.  (with p4runtime_annotation -> p4runtime_translation substitutions)